shapely/constructive.py

```diff
@@ -914,6 +925,21 @@ def snap(geometry, reference, tolerance, **kwargs):
     <POLYGON ((0 0, 0 10, 8 10, 10 0, 0 0))>
     >>> snap(polygon, LineString([(8, 10), (8, 0)]), tolerance=5)
     <POLYGON ((0 0, 0 10, 8 10, 8 0, 0 0))>
+    >>> # example from: https://postgis.net/docs/ST_Snap.html
```

Comment:
What is the intent of the additional examples here?  I'm finding the MultiPolygon examples here a bit hard to follow and not clarifying behavior, which means we should seek out simpler examples unless there is a point you are trying to illustrate regarding MultiPolygons specifically.

I suggest dropping the MultiPolygon and LineString examples added here, and add a simpler example of snapping a line to line (demonstrates cleaning imprecise coordinates): 

```Python
>>> line1 = LineString([(0.1, 0.1), (0.49, 0.51), (1.01, 0.89)])
>>> line2 = LineString([(0, 0), (0.5, 0.5), (1.0, 1.0)])
>>> snap(line1, line2, 0.25)
<LINESTRING (0 0, 0.5 0.5, 1 1)>
```

We also occasionally like to demonstrate behavior for `None` or empty geometries if they lead to unexpected behavior.

For example, I'd suggest adding:

```Python
>>> snap(line1, None, 0.25) is None
True
```

Because without noting it here, a user might instead expect to get back the input geometry (snap to nothing => no snapping).


Likewise, though perhaps more obvious, is behavior for an empty geometry:

```Python
>>> snap(line1, LineString([]), 0.25)
<LINESTRING (0.1 0.1, 0.49 0.51, 1.01 0.89)>
```

Might also be good to demonstrate a very simple 3D example:

```Python
>>> point1 = Point(0.1, 0.1, 0.5)
>>> multipoint = MultiPoint([(0, 0, 1), (0, 0, 0)])
>>> snap(point1, multipoint, 1)
<POINT Z (0 0 1)>
```

I'm not sure if this is deterministic across GEOS versions, for me it picks the first point that is within tolerance.

This might be now getting a bit carried away, but here is an example that shows how snapping only updates one of the coordinates in a Polygon even though all points in the polygon are equidistant, because this would cause the polygon to collapse:

```Python
>>> poly = box(0, 0, 1, 1)
>>> snap(poly, Point(0.5, 0.5), 1)
<POLYGON ((0.5 0.5, 1 1, 0 1, 0 0, 0.5 0.5))>
```



I also discovered an unexpected promotion of dimension that we should probably track down (don't use as example, just demonstrating it here):

```Python
>>> snap(LineString([]), LineString([]), 0.25)
<LINESTRING Z EMPTY>
```

Reply:
@brendan-ward Thanks very much. I've added your examples.

> I suggest dropping the MultiPolygon and LineString examples added here
I kept the `MultiPolygon` example because there is nice graphics for it in JTS.

>I also discovered an unexpected promotion of dimension that we should probably track down (don't use as example, just demonstrating it here):
https://github.com/shapely/shapely/issues/1686

Reply:
Thanks for the updates!

I still think including the PostGIS example is too much for the examples here; they are not easy to follow without the graphic.  We definitely need more illustrative graphics, but that is out of scope here.

Instead, one option would be to put the link in the descriptive part of the docstring instead, and drop the multipolygon example (I'm not keen on this option though).

```Python
... if they would lead to an invalid geometry.

Also see https://postgis.net/docs/ST_Snap.html for additional examples.

Parameters
-----------
```

However, it should be noted that there does not appear to be an existing precedent for linking to PostGIS for examples, and the examples there are in SQL wrapped WKT, and may not be as easily accessible to Shapely users (the visuals yes, the SQL less so).

At the moment, I think it is just fine to leave out both the example and the crosslink to a different toolkit, and leave it as a follow-up to generate graphics in the docs via `shapely.plotting` rather than in the examples here.  I think the examples should be easy to follow just from looking at the code, and we have some work to do to simplify them further across the board (also out of scope here).

Reply:
@brendan-ward thanks for your review!
Regarding the JTS/PostGIS examples:

> the visuals yes, the SQL less so

I also think that the visuals help and the SQL is hard to follow for a shapely user, that's why I've translated the examples from SQL to shapely, so a shapely user can follow the graphics with the shapely examples.
Although I think they are helpful - if you are not inclined to include them I don't mind removing them or leaving this discussion for a follow-up PR as I agree this is already out of scope of this PR.